cmd/crypto/vault_test.go

```diff
@@ -0,0 +1,91 @@
+// Minio Cloud Storage, (C) 2019 Minio, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crypto
+
+import "testing"
+
+var verifyVaultConfigTests = []struct {
+	Config     VaultConfig
+	ShouldFail bool
+}{
+	{
+		ShouldFail: false, // 0
+		Config:     VaultConfig{},
+	},
+	{
+		ShouldFail: true,
+		Config:     VaultConfig{Endpoint: "https://127.0.0.1:8080"},
+	},
+	{
+		ShouldFail: true, // 1
+		Config: VaultConfig{
+			Endpoint: "https://127.0.0.1:8080",
+			Auth:     VaultAuth{Type: "unsupported"},
+		},
+	},
+	{
+		ShouldFail: true, // 2
+		Config: VaultConfig{
+			Endpoint: "https://127.0.0.1:8080",
+			Auth: VaultAuth{
+				Type:    "approle",
+				AppRole: VaultAppRole{},
+			},
+		},
+	},
+	{
+		ShouldFail: true, // 3
+		Config: VaultConfig{
+			Endpoint: "https://127.0.0.1:8080",
+			Auth: VaultAuth{
+				Type:    "approle",
+				AppRole: VaultAppRole{ID: "123456"},
+			},
+		},
+	},
+	{
+		ShouldFail: true, // 4
+		Config: VaultConfig{
+			Endpoint: "https://127.0.0.1:8080",
+			Auth: VaultAuth{
+				Type:    "approle",
+				AppRole: VaultAppRole{ID: "123456", Secret: "abcdef"},
+			},
+		},
+	},
+	{
+		ShouldFail: true, // 5
+		Config: VaultConfig{
+			Endpoint: "https://127.0.0.1:8080",
+			Auth: VaultAuth{
+				Type:    "approle",
+				AppRole: VaultAppRole{ID: "123456", Secret: "abcdef"},
+			},
+			Key: VaultKey{Name: "default-key", Version: -1},
+		},
+	},
+}
+
+func TestVerifyVaultConfig(t *testing.T) {
+	for i, test := range verifyVaultConfigTests {
+		err := test.Config.Verify()
+		if test.ShouldFail && err == nil {
```

Comment:
Why not start using t.Run style tests ?

Reply:
Not able to understand... Are you asking for creating a closure for each test case and calling it with a name?! What is the diff/adv.?!
AFAIK `testing.T.Run(...)` is there to execute the *same* tests for different environments. For example if you have test vectors for a hash function and you have different assembler / generic implementations.

Reply:
See e.g.: https://github.com/minio/highwayhash/blob/master/highwayhash_test.go#L15

Reply:
That is not the only reason it's there, the advantage is that we know which test failed, t.Run("myTest%d") , it allows for test run v/s failed to add more context.

Reply:
Not only that we can also run these tests in parallel.

Reply:
> t.Run("myTest%d") , it allows for test run v/s failed to add more context.

This information is already there. Not able to see the difference between `t.Error(<context>, <cause>)` and `t.Run(<context>, <test-func>)` 

> Not only that we can also run these tests in parallel.

I'm aware of that. But this is only beneficial for large/expensive tests. Note that `Run` does a lot of heavy lifting (allocates a separate call stack, channels for communication a.s.o.) I suspect there is speed advantage for small/cheap tests - like this one.

Reply:
My intention was not that let me show you what I meant @aead - the advantage of t.Run() is quite interesting than just some execution environment it allows us to hand pick tests from a table of tests. Such as this

This is the regular run with this PR

```
~ go test -v -run TestVerifyVaultConfig
=== RUN   TestVerifyVaultConfig
--- PASS: TestVerifyVaultConfig (0.00s)
PASS
ok  	github.com/minio/minio/cmd/crypto	0.007s
```

Now with t.Run changes
```
go test -v -run TestVerifyVaultConfig
=== RUN   TestVerifyVaultConfig
=== RUN   TestVerifyVaultConfig/Test0
=== RUN   TestVerifyVaultConfig/Test1
=== RUN   TestVerifyVaultConfig/Test2
=== RUN   TestVerifyVaultConfig/Test3
=== RUN   TestVerifyVaultConfig/Test4
=== RUN   TestVerifyVaultConfig/Test5
=== RUN   TestVerifyVaultConfig/Test6
--- PASS: TestVerifyVaultConfig (0.00s)
    --- PASS: TestVerifyVaultConfig/Test0 (0.00s)
    --- PASS: TestVerifyVaultConfig/Test1 (0.00s)
    --- PASS: TestVerifyVaultConfig/Test2 (0.00s)
    --- PASS: TestVerifyVaultConfig/Test3 (0.00s)
    --- PASS: TestVerifyVaultConfig/Test4 (0.00s)
    --- PASS: TestVerifyVaultConfig/Test5 (0.00s)
    --- PASS: TestVerifyVaultConfig/Test6 (0.00s)
PASS
ok  	github.com/minio/minio/cmd/crypto	0.007s
```

Let's say I want to only debug `Test2` run nothing else
```
go test -v -run TestVerifyVaultConfig/Test2
=== RUN   TestVerifyVaultConfig
=== RUN   TestVerifyVaultConfig/Test2
--- PASS: TestVerifyVaultConfig (0.00s)
    --- PASS: TestVerifyVaultConfig/Test2 (0.00s)
PASS
ok  	github.com/minio/minio/cmd/crypto	0.006s
```

This is quite advantageous this is not possible in the current style, we have to manually disable tests and its quite cumbersome. So I would suggest we follow this style from now on all our tests.

Reply:
Yeah, @harshavardhana that I'm aware of. I'm not a huge fan of that since it leads to the case where a code change breaks a test, then dev keeps fixing until `go test -run TestYada/Yada1` but additionally should make sure that the fixing does not break another test of `TestYada`. So in the end you run the same tests and in best case avoid running some test cases locally but in worst case catch that during CI.

Summarized, during the change-test loop all tests for the function should, in general, run and not just one case because you could have broken another one... Additionally it is quite nosy and clutters the CLI.

However, we can make that the default. But if so, can we make that official among the dev team?! I just want to avoid bike-shading about test styling depending on reviewers...

Reply:
I wanted to follow this model for some time now because of the advantages stated above, and perhaps we should follow it and as per review be asked such that we add t.Run style.